src/vm/memory/segments.zig

```diff
@@ -168,12 +168,44 @@ pub const MemorySegmentManager = struct {
     /// # Returns
     ///
     /// An `AutoArrayHashMap` representing the computed effective sizes of memory segments.
-    pub fn computeEffectiveSize(self: *Self) !std.AutoArrayHashMap(u32, u32) {
-        for (self.memory.data.items, 0..) |segment, i| {
+    pub fn computeEffectiveSize(self: *Self, allow_tmp_segments: bool) !std.AutoArrayHashMap(i64, u32) {
+        if (self.segment_used_sizes.count() != 0) {
+            return self.segment_used_sizes;
+        }
+
+        // TODO: Check if memory is frozen. At the time of writting this function memory cannot be frozen so we cannot check if it frozen.
+        var first_segment_index: i64 = undefined;
+
+        if (allow_tmp_segments) {
+            first_segment_index = -@as(i64, self.memory.num_temp_segments);
+        } else {
+            first_segment_index = 0;
+        }
+
+        var first_index = first_segment_index;
+        while (first_index < self.memory.num_segments) {
+            try self.segment_used_sizes.put(first_segment_index, 0);
+            first_index += 1;
+        }
+
+        var temp1: i64 = 0;
+        for (self.memory.data.items) |segment| {
```

Comment:
here there is no need for the `temp1` variable. The following is equivalent:

```zig
for (self.memory.data.items, 0..) |segment, i| {
...
}
```

Reply:
actually i tried doing that previously but for loop didn't support `i64`, but after the refactor don't need to loop over negatives values since they are separate arrays. so will update this.